serializers/src/main/java/io/pravega/schemaregistry/serializers/MultiFormatSerializerFactory.java

```diff
@@ -0,0 +1,199 @@
+import io.pravega.client.stream.Serializer;
+import io.pravega.schemaregistry.client.SchemaRegistryClient;
+import io.pravega.schemaregistry.client.SchemaRegistryClientFactory;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.SerializationFormat;
+import io.pravega.schemaregistry.schemas.AvroSchema;
+import io.pravega.schemaregistry.schemas.JSONSchema;
+import io.pravega.schemaregistry.schemas.ProtobufSchema;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static io.pravega.schemaregistry.serializers.SerializerFactoryHelper.*;
+import static io.pravega.schemaregistry.serializers.WithSchema.NO_TRANSFORM;
+
+/**
+ * Internal Factory class for multi format serializers and deserializers. 
+ * These serializers can be used to work with streams when either you dont know the format beforehand or the stream allows
+ * for multiple formats. 
+ */
+@Slf4j
+public class MultiFormatSerializerFactory {
+    // region multi format 
+    static Serializer<WithSchema<Object>> serializer(SerializerConfig config) {
+        return serializerInternal(config, Collections.emptyMap());
+    }
+
+    static Serializer<WithSchema<Object>> deserializerWithSchema(SerializerConfig config) {
+        return deserializerInternal(config, Collections.emptyMap(), NO_TRANSFORM);
+    }
+    
+    /**
+     * A deserializer that can read data where each event could be written with different serialization formats. 
+     * Formats supported are protobuf, avro and json. 
+     * An event serialized with avro is deserialized into {@link GenericRecord}.
+     * An event serialized with protobuf is deserialized into {@link DynamicMessage}.
+     * An event serialized with json is deserialized into {@link java.util.LinkedHashMap}.
+     * 
+     * This also takes a transform function which is applied on the deserialized object and should transform the object 
+     * into the type T.  
+     *
+     * @param config serializer config
+     * @param transform a transform function that transforms the deserialized object based on the serialization format 
+     *                  into an object of type T. 
+     * @param <T> Type of object to get back from deserializer. 
+     * @return a deserializer that can deserialize protobuf, json or avro events into java objects.
+     */
+    static <T> Serializer<T> deserializeAsT(SerializerConfig config,
+                                            BiFunction<SerializationFormat, Object, T> transform) {
+        return deserializeAsTInternal(config, Collections.emptyMap(), transform);
+    }
+    // endregion
+
+    private static Serializer<WithSchema<Object>> serializerInternal(SerializerConfig config,
```

Comment:
And here

Reply:
https://github.com/pravega/schema-registry/pull/41#discussion_r449404120